lib/report.js

```diff
@@ -158,7 +160,123 @@ class ConciseReporter extends Reporter {
   }
 }
 
+class TapReporter extends Reporter {
+  constructor(options) {
+    super(options);
+    this.successful = 0;
+    this.failed = 0;
+    this.skipped = 0;
+    this.failedTests = [];
+    this.testNumeration = 1;
+    this.message = '';
+    this.got = [];
+    this.expected = [];
+  }
+
+  error(test, error) {
+    // remove one success because 'error' may only be received after end
+    // though the test may have already failed so we need a way to check
+    // if we should do this, i.e. store map of successful tests
+    // this.successes--;
+    // test.todo ? this.skipped++ : this.failures++;
+    // TODO(lundibundi): add test.metadata.filename here too once available
+    this.listFailure(
+      test,
+      { message: `Failure after end: ${error ? error.message : 'no message'}` },
+      ` Test failed: ${test.caption}`
+    );
+  }
+
+  parseTestResults(test, offset = 0) {
+    test.results.forEach((res, index) => {
+      const testId = index + 1;
+      if (res.type === 'subtest') {
+        const t = res.test;
+        console.log(`${pad(offset + 2)}# Subtest: ${t.caption}`);
+        this.parseTestResults(t, offset + 2);
+        const prefix = t.success ? 'ok' : 'not ok';
+        console.log(`${pad(offset + 2)}${prefix} ${testId} - ${t.caption}`);
+      } else {
+        res.success
+          ? this.checkSuccess(res, testId, offset + 2)
+          : this.listFailure(res, testId, offset + 2);
+      }
+      this.message = '';
+    });
+    console.log(`${pad(offset + 2)}1..${test.results.length}`);
+    this.testNumeration++;
+  }
+
+  checkSuccess(test, id, offset) {
+    const testPrefix = test.success ? 'ok' : 'not ok';
+    let successfulTest = `${pad(offset)}${testPrefix} ${id}`;
+    if (test.caption !== undefined) successfulTest += ` - ${test.caption}`;
+    console.log(successfulTest);
+  }
+
+  listFailure(test, id, offset) {
+    this.message += `${pad(offset)}not ok ${id}\n`;
+    this.message += `${pad(offset + 2)}---\n`;
+    if (test.message)
+      this.message += `${pad(offset + 2)}message: '${test.message}'\n`;
+    this.message += `${pad(offset + 2)}severity: `;
+    this.message += test.severity ? `${test.severity}\n` : 'fail\n';
+    this.message += `${pad(offset + 2)}data:\n`;
+    const [expected, actual] = [test.expected, test.actual].map(test =>
```

Comment:
I don't see a point in converting results to string here since `.toString()` will be called when you append this to `this.message` several lines below.